### ISSUE TITLE
Fail в тестировании

### ISSUE THREAD

**Issue by pavelpower** [bem/bem-forum-content-ru#470]
Было хотели прикрутить [enb-bem-specs](https://ru.bem.info/tools/bem/enb-bem-specs/readme/)

В общем полная беда.
Templates только на bemhtml, а мы пользуемся BH - ибо он по мнению команды понятнее.

Потом, как оказалось он тянет за собой блок `jquery` - которого у нас просто нет, а еще хочет `i-bem`. 

![i-bem](http://dl2.joxi.net/drive/0002/2530/190946/150630/8bd993b1c2.jpg)

В общем очень странные желания для сборки тестов и не понятно почему ему нужен так `jquery`, и `i-bem`? 

Итог выглядит печально: 

![](http://dl2.joxi.net/drive/0002/2530/190946/150630/6b0c6c807b.jpg) 

Так может кто уже получил опыт заведения unit-тестирования без i-bem, jquery?

Видится это примерно так:

В блоках файлы *.spec.js

В которых написаны тесты в BDD стиле.

Сборка происходит в отдельную папочку, и для сборки используются только те слои переопределения кои используются для development мода.

**Comment by pavelpower** [comment]
Так там и bemhtml в зависимостях нет 
![](http://dl2.joxi.net/drive/0002/2530/190946/150630/25af8f62cf.jpg)

**Comment by pavelpower** [comment]
А все таки он там есть, внутри либы enb-bemxjst

**Comment by tavriaforever** [comment]
@blond

**Comment by blond** [comment]
@pavelpower, привет!

Извини за долгий ответ. 

На данный момент все действительно так. 

Про возможность использовать кастомный шаблонизатор есть задача — https://github.com/enb-bem/enb-bem-specs/issues/11. 

Я пробовал её решить с помощью введения специального понятия мета-технология, чтобы можно было самому настраивать чем и как собирать нужные файлы, в том числе и шаблонизатор в составе клиенского js — https://github.com/enb-bem/enb-bem-specs/pull/13. Но, кажется, такое решение оказалось сложным для использования. 

Есть мысли, что можно сделать опцию, например, `templateEngine: 'BH'`, которой просто указывать какой шаблонизатор использовать `BEMHTML` или `BH`. Такое решение устроит?

По поводу `i-bem.js`, кажется, что он тянется как раз из-за BEMHTML и если поддержать сборку с помощью BH то он приезжать не будет.

Блок `jquery` требуется в блоке `speс`: https://github.com/narqo/bem-pr/blob/master/spec.blocks/spec/spec.deps.js#L19. Раньше это было нужно для корректной работы `mocha` но, кажется, что это давно [уже не так](https://github.com/mochajs/mocha/blob/f08b789e2ca431d23ab46a60a4ec0f8d524e3485/HISTORY.md#0141--2012-03-06), и его можно безболезненно оторвать.

**Comment by pavelpower** [comment]
Салют, @blond! Спасибо за емкий ответ.

Решение по использованию шаблонизатора при помощи опции `templateEngine` на текущий момент выглядит хорошо. 

И если есть возможность выпилить jquery - то это очень хорошо. Т.к. при формировании тестов для проверки на устройствах jquery печальный рудимент.

Вопрос в том, на сколько быстро можно внести эти изменения? 
И как можно помочь, сможешь ли скоорденировать действия поставив задачи?

Это я к чему, инструмент  нужен и уже сейчас, и я готов выделить ресурсы на то, чтобы допилить его. Но так же не хочется терять за зря время и выполнить все скоорденировано.

**Comment by blond** [comment]
> Вопрос в том, на сколько быстро можно внести эти изменения? 
И как можно помочь, сможешь ли скоорденировать действия поставив задачи?

Самый быстрый способ — присылай PR'ы :) 

На все вопросы буду стараться быстро отвечать.

**Comment by blond** [comment]
Пока что завёл задачи: https://github.com/enb-bem/enb-bem-specs/issues/33 https://github.com/enb-bem/enb-bem-specs/issues/34

**Comment by pavelpower** [comment]
@blond Ты только скажи откуда берется `<script class="spec-runner">(function() {var global = this;typeof modules === "object"?  modules.require(["jquery", "mocha", "spec"], function($, mocha) {    init($, mocha);  }) :   init(global.jQuery, global.mocha);function init($, mocha) {  global.mochaPhantomJS? global.mochaPhantomJS.run(done) : mocha.run(done);  function done() { $("#mocha").show() }}}());</script>`

Понять не могу найти не ищется...

**Comment by pavelpower** [comment]
@blond все понял, придется еще и bem-pr править :(

**Comment by pavelpower** [comment]
Хотя там уже есть BH https://github.com/narqo/bem-pr/blob/master/spec.blocks/spec/spec.bh.js

Видимо нужно просто обновить версию

**Comment by blond** [comment]
@pavelpower всё так. Можно, конечно, ещё больше заморочится и попробовать отказаться от `bem-pr`.

Мне не нравится его использование с точки зрения API. Пользователь вынужден подключать `bem-pr` в уровни. Логичнее, чтобы необходимый код был в `enb-bem-specs`.

**Comment by pavelpower** [comment]
@narqo признайся, зачем ты туда BH добавил?

Не уж, то bem-spec свой уже с BH сделал?

**Comment by pavelpower** [comment]
@blond, да, ты прав bem-pr тут лишний однозначно. Лучше допишу блоки внутри `enb-bem-specs`

**Comment by narqo** [comment]
> зачем ты туда BH добавил

Мне для какого-то своего проекта на bh нужно было написать тестов на блоки, чтобы удобнее было разрабатываться. Быстрее оказалось взять уровни из bem-pr, накидать сборку обычного бандла на enb и запускал тесты в браузере. Для разработки подошло, а никакого cli + ci там и не предполагалось. С тех в bem-pr есть bh-шаблоны.

**Comment by pavelpower** [comment]
@narqo Спасибо!

**Comment by pavelpower** [comment]
@blond ты посмотри что получается

https://github.com/pavelpower/enb-bem-specs/blob/master/lib/node-configurator.js

Я смотрю и стили исключительно через `.styl ` используются.
Но так повелось, что у нас в проекте прижился `.less`

И похоже, что нужно менять подход к данной либе, не очень все таки использовать настройки.
Как-то плагинами надо, что-ли.

А какие там у тебя идеи были то?

**Comment by kompolom** [comment]
@pavelpower Наткнулся на те же грабли. Скажи сейчас в каком состоянии идея с поддержкой bh?
\cc @blond

**Comment by tadatuta** [comment]
> Скажи сейчас в каком состоянии идея с поддержкой bh?

`enb-bh` умеет собирать `BH` [с мимикрией](https://ru.bem.info/tools/bem/enb-bh/readme/#Мимикрия-под-bemhtml) под `BEMHTML`, поэтому с точки зрения `enb-bem-specs` разницы вообще не будет.

**Comment by kompolom** [comment]
@tadatuta `enb-bh` умеет, но [тут](https://github.com/enb-bem/enb-bem-specs/blob/master/lib/node-configurator.js) bemhtml подключается. Как ему можно bh скормить?

**Comment by pavelpower** [comment]
@komplomon ну как как решил, запилил свой enb-bem-specs с преферансом и поэтессами

https://github.com/pavelpower/enb-bem-specs

**Comment by pavelpower** [comment]
Теперь правда btmhtml не умеет а bh умеет

**Comment by kompolom** [comment]
@pavelpower Сейчас пробую запустить тесты bem-components c твоим форком enb-bem-specs. Тесты валятся ругаясь на неподключеный BEMTHML.

**Comment by pavelpower** [comment]
@kompolom я делал только под BH, т.к. времени не было на реализацию вместе BH и BEMHTML. 
У нас в проекте шаблоны только BH.

**Comment by pavelpower** [comment]
@kompolom вся магия в файле https://github.com/pavelpower/enb-bem-specs/blob/master/lib/node-configurator.js

в моем форке отличие два:
1. используется less
2. используется BH

**Comment by kompolom** [comment]
@pavelpower Все так, но в bem-components есть BH шаблоны. А less я обратно переключил на stylus

**Comment by pavelpower** [comment]
@kompolom тогда как решишь проблему с "Тесты валятся ругаясь на неподключеный BEMTHML." отпиши

**Comment by kompolom** [comment]
Добавлю, что теперь есть [PR](https://github.com/enb-bem/enb-bem-specs/pull/41) и есть [форк](https://github.com/kompolom/enb-bem-specs/tree/33%40add-template-engine-option), в котором можно выбрать шаблонизатор. Возможно не очень красиво, но работает.

**Comment by pavelpower** [comment]
@kompolom, по ходу забыл ты раскрыть как выбирать в конфигах технологию.

+ Надо бы собратся и продумать универсальненькое решение, когда есть и то и то.

**Comment by kompolom** [comment]
@pavelpower там в examples есть пример
https://github.com/kompolom/enb-bem-specs/blob/33%40add-template-engine-option/examples/silly/.enb/make.js

**Comment by pavelpower** [comment]
@blond может принять? https://github.com/kompolom/enb-bem-specs/
